### PR TITLE
KeyAnalyzer tests.

### PR DESCRIPTION
**Pull request by IDisposable**
This is based extracting the unit tests for _KeyAnalyzer_ and a couple trivial code simplifications without the huge changes in PR #89863

Hoisted the calculation of the maximum collisions allow outside the `HasSufficientUniquenessFactor` as the calculation only needs to be done once (also makes it more testable).

Made `ContainsAnyLetters` and `SufficientUniquenessFactor` `internal` so they can be tested.

Added tests for `ContainsAnyLetters` and `SufficientUniquenessFactor` are working correctly.

### CODE REVIEW

**Comment by IDisposable** [comment]
Benchmarks are identical (no performance difference nor allocation)
```

BenchmarkDotNet v0.13.7-nightly.20230717.35, Windows 11 (10.0.22631.2129)
Intel Core i7-10875H CPU 2.30GHz, 1 CPU, 16 logical and 8 physical cores
.NET SDK 8.0.100-preview.7.23322.33
  [Host]     : .NET 8.0.0 (8.0.23.32106), X64 RyuJIT AVX2
  Job-KUPTKU : .NET 8.0.0 (42.42.42.42424), X64 RyuJIT AVX2
  Job-XTFASN : .NET 8.0.0 (42.42.42.42424), X64 RyuJIT AVX2
  Job-TLRRVP : .NET 8.0.0 (42.42.42.42424), X64 RyuJIT AVX2

PowerPlanMode=00000000-0000-0000-0000-000000000000  Arguments=/p:EnableUnsafeBinaryFormatterSerialization=true  IterationTime=250.0000 ms  
MaxIterationCount=20  MinIterationCount=15  WarmupCount=1  

```


| Method                            | Job            | Toolchain                                                                                                        | Count     |              Mean |            Error |           StdDev |            Median |               Min |               Max |    Ratio | MannWhitney(2ms) |  RatioSD |        Gen0 |        Gen1 |        Gen2 |    Allocated | Alloc Ratio |
| ----------------------------------- | ---------------- | ------------------------------------------------------------------------------------------------------------------ | ----------- | ------------------: | -----------------: | -----------------: | ------------------: | ------------------: | ------------------: | ---------: | ------------------ | ---------: | ------------: | ------------: | ------------: | -------------: | ------------: |
| ToFrozenDictionary                | Job-KUPTKU     | Simple                                                                                                           | 10        |       1,231.76 ns |        23.212 ns |        23.837 ns |       1,237.90 ns |       1,180.80 ns |       1,264.89 ns |     1.00 | Same             |     0.02 |      0.2034 |           - |           - |       1720 B |        1.00 |
| ToFrozenDictionary                | Job-XTFASN     | Baseline                                                                                                         | 10        |       1,227.08 ns |        24.197 ns |        25.890 ns |       1,237.73 ns |       1,180.79 ns |       1,264.45 ns |     1.00 | Base             |     0.00 |      0.2053 |           - |           - |       1720 B |        1.00 |
|                                   |                |                                                                                                                  |           |                   |                  |                  |                   |                   |                   |          |                  |          |             |             |             |              |             |
| ToFrozenDictionary                | Job-KUPTKU     | Simple                                                                                                           | 100       |       7,284.33 ns |        95.393 ns |        89.231 ns |       7,293.21 ns |       6,983.58 ns |       7,367.73 ns |     1.01 | Same             |     0.02 |      1.4291 |           - |           - |      12112 B |        1.00 |
| ToFrozenDictionary                | Job-XTFASN     | Baseline                                                                                                         | 100       |       7,216.35 ns |       108.905 ns |       101.870 ns |       7,257.10 ns |       6,953.81 ns |       7,325.22 ns |     1.00 | Base             |     0.00 |      1.4441 |           - |           - |      12112 B |        1.00 |
|                                   |                |                                                                                                                  |           |                   |                  |                  |                   |                   |                   |          |                  |          |             |             |             |              |             |
| ToFrozenDictionary                | Job-KUPTKU     | Simple                                                                                                           | 1000      |      62,515.42 ns |     1,211.874 ns |     1,074.295 ns |      62,012.86 ns |      61,724.80 ns |      64,788.34 ns |     0.98 | Same             |     0.03 |     11.2245 |      1.2755 |           - |      94864 B |        1.00 |
| ToFrozenDictionary                | Job-XTFASN     | Baseline                                                                                                         | 1000      |      63,883.60 ns |     1,310.970 ns |     1,509.715 ns |      63,058.56 ns |      61,999.56 ns |      66,311.95 ns |     1.00 | Base             |     0.00 |     11.2448 |      1.3075 |           - |      94864 B |        1.00 |
|                                   |                |                                                                                                                  |           |                   |                  |                  |                   |                   |                   |          |                  |          |             |             |             |              |             |
| ToFrozenDictionary                | Job-KUPTKU     | Simple                                                                                                           | 10000     |   1,159,184.94 ns |    22,976.312 ns |    21,492.056 ns |   1,160,790.62 ns |   1,112,379.46 ns |   1,190,017.86 ns |     1.00 | Same             |     0.01 |    147.3214 |    147.3214 |    147.3214 |     926134 B |        1.00 |
| ToFrozenDictionary                | Job-XTFASN     | Baseline                                                                                                         | 10000     |   1,154,967.58 ns |    16,610.134 ns |    13,870.214 ns |   1,159,534.82 ns |   1,124,604.91 ns |   1,170,523.21 ns |     1.00 | Base             |     0.00 |    147.3214 |    147.3214 |    147.3214 |     926134 B |        1.00 |

**Comment by stephentoub** [comment]
> Benchmarks are identical

Can we please undo the hoisting-related changes then?  Let's keep this just tests.

**Comment by IDisposable** [comment]
Will do :)

File: src/libraries/System.Collections.Immutable/tests/Frozen/KeyAnalyzerTests.cs [hunk 1]
```diff
@@ -212,5 +214,75 @@ public static void IsAllAscii()
...
+            HashSet<string> set = new HashSet<string>(StringComparer.Ordinal);
+
+            Assert.True(KeyAnalyzer.HasSufficientUniquenessFactor(set, new[] { "a", "b", "c" }, 0));
+            Assert.Equal(3, set.Count);
+
+            Assert.True(KeyAnalyzer.HasSufficientUniquenessFactor(set, new[] { "a", "b", "a" }, 1));
+            Assert.Equal(2, set.Count); // set should only have the non-collided ones
+
+            Assert.False(KeyAnalyzer.HasSufficientUniquenessFactor(set, new[] { "aa", "ab", "aa" }, 0));
+            Assert.Equal(0, set.Count); // if we failed it should empty the set
```
Comment (adamsitnik) [inline src/libraries/System.Collections.Immutable/tests/Frozen/KeyAnalyzerTests.cs:240]: This is not true and that is why the test is failing.

```suggestion
            Assert.Equal(2, set.Count);
```

**Review by adamsitnik — COMMENTED** [review]
Thank you for your contribution!

The test is failing, but I think I've found the reason. I am going to apply my suggestion and merge it if the CI gets green.

**Comment by IDisposable** [comment]
Yep, that was from when it was still using my other `KeyAnalyzer` changes (that resulted in an empty set when `HasSufficientUniquenessFactor` returned `false`). 

Since we're not using that code right now, the test was wrong, sorry.

Thanks for fixing it.

**Review by adamsitnik — APPROVED** [review]
LGTM, thank you @IDisposable !

**Pull request merged**